### PR TITLE
Custom logos read from files

### PR DESCRIPTION
**Pull request by alborrajo**
If the logo specified with ```--logo``` isn't recognized, the program will check if it's a file that can be read, and if it is, the contents of the file will be used. Otherwise it'll use the unknown logo.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 fastfetch is a [neofetch](https://github.com/dylanaraps/neofetch) like tool for fetching system information and displaying them in a pretty way. It is written in c to achieve much better performance, in return only linux is supported. It also uses mechanisms like multithreading and caching to finish as fast as possible.  
   
-The output on my personal machine with default configurations looks like this:
-<img src="screenshots/example.png" alt="example output" width="100%"/>  
+This fork adds custom logos read from files:
+<img src="screenshots/amogus.png" alt="example output" width="100%"/>
```
Comment (LinusDierheimer) [inline README.md:6]: This changes are fine for your fork but shouldn't be in this repository.

File: src/common/logo.c [hunk 2]
```diff
@@ -249,6 +269,37 @@ static void loadGarudaLogo(FFlogo* logo, bool doColor)
     sprintf(logo->chars[19], FASTFETCH_TEXT_MODIFIER_BOLT"%s                                              "FASTFETCH_TEXT_MODIFIER_RESET, color);
 }
 
+static void loadLogoFromFile(FFlogo* logo, bool doColor, char* pathToFile)
+{
+    strcpy(logo->name, "amogus");
+
+    const char* color = doColor ? "\033[91m" : "";
+    strcpy(logo->color, "\033[91m");
```
Comment (LinusDierheimer) [inline src/common/logo.c:277]: I don't think we should set the same color for every logo loaded from file. Use the `--color` option to set it.

File: src/common/logo.c [hunk 3]
```diff
@@ -249,6 +269,37 @@ static void loadGarudaLogo(FFlogo* logo, bool doColor)
     sprintf(logo->chars[19], FASTFETCH_TEXT_MODIFIER_BOLT"%s                                              "FASTFETCH_TEXT_MODIFIER_RESET, color);
 }
 
+static void loadLogoFromFile(FFlogo* logo, bool doColor, char* pathToFile)
+{
+    strcpy(logo->name, "amogus");
```
Comment (LinusDierheimer) [inline src/common/logo.c:274]: No need to do this, as we never load the logo by name. Maybe just set a null byte at index 0 to be save.

File: src/common/logo.c [hunk 4]
```diff
@@ -249,6 +269,37 @@ static void loadGarudaLogo(FFlogo* logo, bool doColor)
...
+        fgets(buff, sizeof(buff), fp);
+        normalizedBuff = strtok(buff, "\n");
+        
+        sprintf(logo->chars[logoHeight++],   FASTFETCH_TEXT_MODIFIER_BOLT"%s%s"FASTFETCH_TEXT_MODIFIER_RESET, color, normalizedBuff);
+
+        len = utf8_strlen(normalizedBuff);
+        logoWidth = len > logoWidth
+            ? len
+            : logoWidth;
+    }
```
Comment (LinusDierheimer) [inline src/common/logo.c:295]: Wouldn't it be simpler to use `getline`?

File: src/common/logo.c [hunk 5]
```diff
@@ -272,6 +323,8 @@ void ffLoadLogoSet(FFconfig* config, const char* logo)
         loadGarudaLogo(&config->logo, config->colorLogo);
     else if(strcasecmp(logo, "unknown") == 0)
         loadUnknownLogo(&config->logo);
+    else if(access(logo, R_OK) == 0)
+        loadLogoFromFile(&config->logo, config->colorLogo, logo);
```
Comment (LinusDierheimer) [inline src/common/logo.c:327]: I think we should also introduce an extra option for this. Maybe `--logo-file`? This way the user has better control about what logo is loaded. Anyway we can leave this a a fallback for `--logo`.

File: src/common/logo.c [hunk 6]
```diff
@@ -249,6 +269,37 @@ static void loadGarudaLogo(FFlogo* logo, bool doColor)
...
+    int len;
+
+    int logoWidth = 0, logoHeight = 0;
+    while(!feof(fp)) {
+        fgets(buff, sizeof(buff), fp);
+        normalizedBuff = strtok(buff, "\n");
+        
+        sprintf(logo->chars[logoHeight++],   FASTFETCH_TEXT_MODIFIER_BOLT"%s%s"FASTFETCH_TEXT_MODIFIER_RESET, color, normalizedBuff);
+
+        len = utf8_strlen(normalizedBuff);
```
Comment (LinusDierheimer) [inline src/common/logo.c:291]: Im not into this topic, why isn't a simple `strlen` enough? The point of Ascii logos is that they can be encoded using ascii so i'm not sure if we need to support unicode.

**Comment by alborrajo** [comment]
I'll fix all of this when I get some time later. Sorry if some things don't
make much sense, I'm pretty much inexperienced in both C development and
doing pull request.
The logo name being amogus was something I forgot to remove.
strlen gave me wrong results when trying to load logos that included UTF-8
characters, so I added that function to support them.
I didn't know about getline, as I said I barely have experience with C.
I'll take a look at it.

File: src/common/logo.c (hunk 2 above)
Comment (alborrajo) [reply src/common/logo.c:277]: I'm having trouble getting color to work even when using the examples in ```fastfetch --help color```. None of them seem to work, it just prints the color encoding sequence as it is, even on the information lines (OS, host, kernel, etc)

File: src/common/logo.c (hunk 4 above)
Comment (alborrajo) [reply src/common/logo.c:295]: I don't see how using getline would make it simpler, getline gets the same data as fgets, I still have to strip the newline character from the read data

File: src/common/logo.c (hunk 2 above)
Comment (LinusDierheimer) [reply src/common/logo.c:277]: This actually was a bug, fixed in the latest commit.

File: src/common/logo.c (hunk 4 above)
Comment (LinusDierheimer) [reply src/common/logo.c]: `getline` returns the number of bytes read, so removing the 'ņ' is faster.

File: src/common/logo.c (hunk 5 above)
Comment (LinusDierheimer) [reply src/common/logo.c]: Just like all flags are handled: The last one is used. So you can overwrite previous options.

File: src/common/logo.c (hunk 6 above)
Comment (alborrajo) [reply src/common/logo.c]: It's not necessary, but it's a nice bonus. If you don't think there has to be support for UTF8 i can just replace it with strlen

File: completions/bash [hunk 7]
```diff
@@ -192,6 +192,7 @@ __fastfetch_completion()
     local FF_OPTIONS_LOGO=(
         "-l"
         "--logo"
+        "--logo-file"
```
Comment (LinusDierheimer) [inline completions/bash:195]: This is a path option, not a logo option. Logo options are completed with the names of the build in logos.

File: src/common/logo.c [hunk 8]
```diff
@@ -249,29 +269,71 @@ static void loadGarudaLogo(FFlogo* logo, bool doColor)
...
+            buff[lineBytes-1] = '\0';
+        
+        sprintf(logo->chars[logoHeight++],   FASTFETCH_TEXT_MODIFIER_BOLT"%s%s"FASTFETCH_TEXT_MODIFIER_RESET, color, buff);
+
+        lineLength = utf8_strlen(buff);
+        logoWidth = lineLength > logoWidth
+            ? lineLength
+            : logoWidth;
+    }
+
```
Comment (LinusDierheimer) [inline src/common/logo.c:296]: insert a 
```c
if(buff != NULL)
    free(buff);
```
here

**Comment by LinusDierheimer** [comment]
Thanks for the contribution!

**Comment by LinusDierheimer** [comment]
It doesn't load the OS logo anymore on my machine, but i will fix that by my self. You did great work providing this code to build upon.

**Comment by alborrajo** [comment]
Yes, in my machine both work as it should

**Pull request merged**